Testing

katonaaron-msg opened the pull request:
- Created unit tests for CSVConverterUtil, SingleLocationSelectionAlgorithm and MostAbundantLocationSelectionAlgorithm
- Created an integration test for OrderService
- Created Postman API tests for -- order creation: successful, missing stock, unknown product id -- export stock

- Created a generic IntegrationTest superclass
- Created a test for test containers
- Created a controller for clearing and populating the database

katonaaron-msg committed:
Testing

- Created unit tests for CSVConverterUtil, SingleLocationSelectionAlgorithm and MostAbundantLocationSelectionAlgorithm
- Created an integration test for OrderService
- Created Postman API tests for
-- order creation: successful, missing stock, unknown product id
-- export stock

- Created a generic IntegrationTest superclass
- Created a test for test containers
- Created a controller for clearing and populating the database

cristinaAnca commented on src/main/java/ro/msg/learning/shop/DummyData.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.stream.Collectors;
+
+public final class DummyData {
+    public static final Address address1;
```
Please use a suggestive naming on the dummy objects (address, products, location etc)

katonaaron-msg replied:
I renamed address1 to address. As for the products, I created two of them, but I think the best way to differentiate between them is to append a number to their names (e.g. product1, product2 -> two different products, their fields are irrelevant). As for the locations I defined "location20" to be a location having all products in stock, each with a quantity of 20. I renamed it to "locationWithStock20" for better readability.

cristinaAnca replied:
Why do you say that the fields are irrelevant? When you create tests for your application, you should use realistic data, no matter what. In this way you make sure that your application is properly tested. Instead of product1 and product2 create 2 realistic products (for example book and pencil). And for address, create a realistic address with a proper streetName, number, city and so on. In this way you can name the address object using the city name for example, or whatever you want to describe the address.

katonaaron-msg replied:
renamed them

cristinaAnca commented on src/main/java/ro/msg/learning/shop/controller/TestController.java:
```diff
@@ -0,0 +1,54 @@
+public class TestController {
+    private final Flyway flyway;
+    @Autowired
+    private ProductRepository productRepository;
```
In a controller, you should use the services, not the repositories. The Repositories are part of Data Access layer in a layered architecture and the Controllers are part of Presentation layer. The Presentation layer should not directly communicate with the Data Access Layer, they should communicate using the Business Logic Layer (which means the Services). 
![image](https://github.com/ro-msg-spring-training/online-shop-katonaaron-msg/assets/127387393/2363ccae-8567-4d9a-964a-c2634fb78fa4)

katonaaron-msg replied:
I created an IntegrationTestDBService.

cristinaAnca commented on src/test/java/ro/msg/learning/shop/controller/messageconverter/CSVConverterUtilTest.java:
```diff
@@ -0,0 +1,68 @@
+
+class CSVConverterUtilTest {
+
+    private final Person person1 = new Person("John Smith", 21, "johnsmith@gmail.com");
```
Adjust the naming here

katonaaron-msg replied:
How should I name it?

cristinaAnca replied:
john, or johnSmith. And the person2 you can rename it to maria.

katonaaron-msg replied:
renamed them

cristinaAnca commented on src/test/java/ro/msg/learning/shop/service/OrderServiceIntegrationTest.java:
```diff
@@ -0,0 +1,97 @@
+public class OrderServiceIntegrationTest extends IntegrationTest {
+
+    @Autowired
+    private OrderService orderService;
```
You already have the @Before method in the IntegrationTest which populate the DB, why do you populate again in the tests?

katonaaron-msg replied:
In the IntegrationTest superclass I create only the schema. The tables are not populated. So each test in the @before method can choose which data to save in the database before executing the tests. This way the tests are better isolated from external changes such as the flyway migration scripts and the content of the database is easier to understand while reading the code. Moreover, the integration tests may prepare the database in different ways, therefore this population is not done in the superclass.

cristinaAnca replied:
I agree with you, and I think I misunderstood what you are doing in the superclass

katonaaron-msg replied:
removed it

cristinaAnca commented on src/test/java/ro/msg/learning/shop/service/OrderServiceIntegrationTest.java:
```diff
@@ -0,0 +1,97 @@
+    @Autowired
+    private OrderService orderService;
+    @Autowired
+    private ProductRepository productRepository;
```
Also, the integration tests should test the entire flow from the controllers to the DB. So in the integration tests you should call the endpoints, not the repositories or services.

katonaaron-msg replied:
I created integration tests for the OrderController.

cristinaAnca replied:
So I think you are not making the difference between an integration test and unit test. If you want to test the service methods, then it is an Unit Test. Integration Tests will test the entire flow from the endpoint (which is in the controller) and the Database Layer. If you create unit test, then you will not use the database. Unit tests will mock the call to the other components which are not part of the "unit" (or component) under testing.

katonaaron-msg replied:
removed

cristinaAnca commented on src/test/java/ro/msg/learning/shop/testcontainer/TestContainerTest.java:
```diff
@@ -0,0 +1,54 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class TestContainerTest extends IntegrationTest {
```
What is this test for?

katonaaron-msg replied:
I renamed it to IsolationTest to reflect on what it does: it verifies that each test is isolated by checking whether the location table is empty, saving a new location, and then in the second test checking that the table is empty again.

cristinaAnca replied:
This is redundant, you are testing that the tests works correctly :D If you want, you can keep it like this, but in my opinion  it makes no sense to test this. With the tests you should test the functionalities of your application.

katonaaron-msg replied:
removed it

cristinaAnca commented on src/test/java/ro/msg/learning/shop/testcontainer/TestContainerTest.java:
```diff
@@ -0,0 +1,54 @@
+public class TestContainerTest extends IntegrationTest {
+
+    @Autowired
+    LocationRepository locationRepository;
```
Again the repositories should not be used in tests

katonaaron-msg replied:
I wanted to persist an entity in the database, so I used a repository. How should I modify it?

cristinaAnca replied:
If it is an integration test (which I see it is because it extends IntegrationTest), then you should call the endpoints using mockMvc.perform(). You already have a controller for testing purposes, you can create an endpoint there which will receive an entity as parameter and persist it in the DB.

katonaaron-msg replied:
removed the class

katonaaron-msg committed:
Testing

- create integration test for OrderController
- rename variables in DummyData
- move logic from TestController to IntegrationTestDBService
- rename TestContainerTest to IsolationTest
- remove unused field in CreateOrderDTO

cristinaAnca commented on src/main/java/ro/msg/learning/shop/DummyData.java:
```diff
@@ -0,0 +1,110 @@
+    public static final Customer customer;
+
+    static {
+        customer = new Customer();
```
Please use here the constructor with parameters and then you will don't need to use the setters anymore. In this way your code is more readable.

katonaaron-msg replied:
constructor created and used

cristinaAnca commented on src/main/java/ro/msg/learning/shop/DummyData.java:
```diff
@@ -0,0 +1,110 @@
+
+    }
+
+    public static Address createAddress(int no) {
```
Please use realistic data as I said before, city1, city2 and so on are not realistic. You can add every property of the Address as parameters to this method and then inside the method you set these parameters to the address you return. Note: use the constructor with parameters and get rid of setters calls.

katonaaron-msg replied:
realistic data used and method removed

cristinaAnca commented on src/main/java/ro/msg/learning/shop/DummyData.java:
```diff
@@ -0,0 +1,110 @@
+        return address;
+    }
+
+    public static Product createProduct(int no) {
```
Same comment as for createAddress method.

katonaaron-msg replied:
removed

cristinaAnca commented on src/main/java/ro/msg/learning/shop/DummyData.java:
```diff
@@ -0,0 +1,110 @@
+        return product;
+    }
+
+    public static Location createLocation(int productQuantity) {
```
Please do a refactoring here in order to have just one reusable method for createLocation. It is confusing to have 4 methods which do the same thing (create a Location) but in different ways. You can have just one method to create a location with multiple parameters.

katonaaron-msg replied:
method removed

cristinaAnca commented on src/test/java/ro/msg/learning/shop/controller/OrderControllerIntegrationTest.java:
```diff
@@ -0,0 +1,137 @@
+    @Autowired
+    private MappingJackson2HttpMessageConverter springMvcJacksonConverter;
+    @Autowired
+    private ProductRepository productRepository;
```
Same comment as before, please get rid of repositories in integration tests and use endpoint calls with mockMvc.perform

katonaaron-msg replied:
repositories are removed and requests to TestController  are added

katonaaron-msg committed:
Refactor tests

- Rename variables containing dummy data
- Delete OrderServiceIntegrationTest and IsolationTest
- Update OrderControllerIntegrationTest to use the TestController for database population
- Use constructors instead of setters
